EndlessClient/Dialogs/GuildDialog.cs

```diff
@@ -1,233 +1,522 @@
+                         ITextMultiInputDialogFactory textMultiInputDialogFactory,
+                         IContentProvider contentProvider,
+                         IGuildActions guildActions,
+                         IGuildSessionProvider guildSessionProvider,
+                         ICharacterInventoryProvider characterInventoryProvider,
+                         IEIFFileProvider eifFileProvider)
             : base(nativeGraphicsManager, dialogButtonService, DialogType.Guild)
         {
             _dialogIconService = dialogIconService;
             _localizedStringFinder = localizedStringFinder;
             _characterProvider = characterProvider;
             _messageBoxFactory = messageBoxFactory;
+            _textMultiInputDialogFactory = textMultiInputDialogFactory;
+            _contentProvider = contentProvider;
+            _guildActions = guildActions;
+            _guildSessionProvider = guildSessionProvider;
+            _characterInventoryProvider = characterInventoryProvider;
+            _eifFileProvider = eifFileProvider;
+            _stateStack = new Stack<State>();
+            _creationMemberCount = 0;
 
-            SetState(GuildDialogState.Initial);
+            SetState(new State(GuildDialogState.Initial));
 
             BackAction += (_, _) =>
             {
-                switch (_state)
+                if (_stateStack.Count > 0)
                 {
-                    case GuildDialogState.Management:
-                        SetState(GuildDialogState.Initial);
-                        break;
-                    default:
-                        // no-op
-                        break;
+                    var previousState = _stateStack.Pop();
+                    SetState(previousState);
+                }
+                else
+                {
+                    SetState(new State(GuildDialogState.Initial));
                 }
             };
 
             Title = _localizedStringFinder.GetString(EOResourceID.GUILD_GUILD_MASTER);
         }
 
-        private void SetState(GuildDialogState state)
+        protected override void OnUpdateControl(GameTime gameTime)
         {
-            if (state != GuildDialogState.Initial && _state == state)
-                return;
+            if (_state.DialogState == GuildDialogState.RegisterGuild)
+            {
+                _guildSessionProvider.CreationSession.MatchSome(creationSession =>
+                {
+                    if (creationSession.Approved)
+                    {
+                        SetState(new State(GuildDialogState.WaitingForMembers), false);
+                    }
+                });
+            }
+
+            if (_state.DialogState == GuildDialogState.WaitingForMembers)
+            {
+                _guildSessionProvider.CreationSession.MatchSome(creationSession =>
+                {
+                    if (creationSession.Members.Count != _creationMemberCount)
```

Comment:
Do members only ever get added to this list? Are they always added one at a time?

I'd prefer an approach that sets `_creationMemberCount` to `creationSession.Members.Count`, and clears/regenerates the list dialog items each time the list is updated. You could also take the more arduous approach of detecting added/removed/updated members but I think clear/recreate is probably good enough for this case.

The current method seems error-prone in it's assumption that members will only ever be added to this list if the counts are different

Reply:
There's no way for a pending creation member to be removed

Reply:
Took another look at the guild reply handler, this should be fine as-is